text/0000-tuple-cons-cell.md

```diff
@@ -0,0 +1,262 @@
+        (a, b; (c,))
+        (a, b,; (c,))
+        (a, b; (c; ()))
+        (a, b; (c,; ()))
+        (a, b,; (c; ()))
+        (a, b,; (c,; ()))
+        (a; (b, c))
+        (a; (b, c,))
+        (a,; (b, c))
+        (a,; (b, c,))
+        (a; (b, c; ()))
+        (a; (b, c,; ()))
+        (a,; (b, c; ()))
+        (a; (b, c,; ()))
+        (a,; (b, c,; ()))
+        (a; (b; (c,)))
+        (a; (b,; (c,)))
+        (a,; (b; (c,)))
+        (a,; (b,; (c,)))
+        (a; (b; (c; ())))
+        (a; (b; (c,; ())))
+        (a; (b,; (c; ())))
+        (a; (b,; (c,; ())))
+        (a,; (b; (c; ())))
+        (a,; (b; (c,; ())))
+        (a,; (b,; (c; ())))
+        (a,; (b,; (c,; ())))
+
+    and so forth...
+
+This RFC proposes equivalent syntax for tuple types. Formally, the syntax for
+tuple types could be described with the following grammar fragment:
+
+```
+ty_tuple
+: "(" ")"
+| "(" ty "," ty_tuple_inner ")"
+| "(" ty_tuple_inner ";" ty ")"
+
+ty_tuple_inner
+: %empty
+| ty
+| ty "," ty_tuple_inner
+```
+
+With this syntax, any tuple can be expressed as either `()` or `(head; tail)`.
+This makes it possible for generic code to handle all tuples by covering just
+these two cases.
+
+In addition to syntax for expressions and types, this RFC also proposes syntax
+for destructuring tuples into a head and tail. Here, the obvious syntax is
+used, ie. `let (head; tail) = (a; b);` results in `head == a` and
+`tail == b`.
+
+This RFC also proposes a new marker trait be added to the language. `Tuple` is
+a trait which is implemented for `()` and `(H; T) where T: Tuple`. In general,
+the type `(H; T)` is only valid when `T: Tuple`.
+
+### Representation
+
+The main problem with implementing this RFC is the question of representation.
+At the memory level, in current Rust, there is no guarantee that the
+representation of an `(a; b)` contains the representation of a `b`. The
+solution proposed here is two-fold. First, we allow types to have separate
+stride and size as per RFC issue #1397. Secondly, we layout tuples in reverse
+order. Under this scheme, the tuple `(A, B, C) : (u16, u16, u32)` would be
+represented as
+
+```
+------------------------------------------------
+| Byte | 0  | 1  | 2  | 3  | 4  | 5  | 6  | 7  |
+|------|-------------------|---------|---------|
+| Data |      C (u32)      | B (u16) | A (u16) |
+------------------------------------------------
+```
+
+And it's tail `(B, C) : (u16, u32)` would be represented as
+
+```
+------------------------------------------------
+| Byte | 0  | 1  | 2  | 3  | 4  | 5  | 6  | 7  |
+|------|-------------------|---------|---------|
+| Data |      C (u32)      | B (u16) | padding |
+------------------------------------------------
+```
+
+Crucially, the stride of this type is only 6 bytes. This means that a
+`&mut (u16, u32)` can not be used to modify the tuple's two trailing "padding"
+bytes as any tuple (accessed through a reference) may be the tail of a larger
+tuple.
+
```

Comment:
> Would this representation permit the last element of a tuple to be unsized? E.g. `&(bool, char, Display)`.

Good point, that's a pretty strong argument for changing the syntax to something like `(head..., tail)` where the subtuple is collected into the head rather than the tail. I'll see what comes of the conversation on #1592 before updating the RFC.

> Would this be backwards compatible with existing `unsafe` code which assumes that all of `size_of::<T>()` can be overwritten?

Well, in general having, a size/stride distinction would break code like that. Unless we make `size_of` actually return the stride for the sake of backwards-compatibility.